# Evaluating damaged cell detection strategies for single cell RNA sequencing (scRNA-seq)

This repository contains the workflow, direct analysis code, and original data for the results and plots generated in our article. For any concerns, comments, questions or suggestions, please open a new  [![Issue](https://img.shields.io/badge/Issues-blue?style=flat&logo=github)](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/issues). 

<br>

A replicate of this repository with the input, and resulting, data and plots is available on ```Zenodo```, [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.14048075.svg)](https://doi.org/10.5281/zenodo.14048075). In downloading and extracting this ```Zenodo``` repository on your local system, we hope to make our analysis results reproducible. However, as it is not containarised, complications may arise in future implementation of the code because of dependency issues, although attempts to include package versions in scripts have been made to address this. 


<br>

⚪[  Analysis Overview](#-analysis-overview) | ⚪[  Data Availability](#-data-availability) ⚪[  Analysis Scripts](#-analysis-scripts) | 

<br>
<br>

## ⚪ Analysis Overview
The analysis is made up of four components beginning with a search for manual and tool-based strategies for damaged cell filtering, A. To test the strategies, data was collected and processed, B, for testing and comparison, C. An overview of the results are condensed, D,  into sections for the user, containing usage recommendations, and the developers, containing areas for improvement. 
<br>
![Alt text](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/images/study_workflow.png)

<br>

## ⚪ Data Availability   

All data used for the analysis and generated by the analysis is publicly available online at the links provided in the table below. 
<br>

<br>

#### Ground truth data | 2 datasets, 5 damaged samples (*)

|   | Dataset           | Treatment                          | Sorting                           | Sample      | Accession | Download |
|---|-------------------|------------------------------------|-----------------------------------|-------------|-----------|----------|
| | | | | | |||
| * | HEK293 cell line  | Staurosporine (1 μM, 2 hours)      | Annexin V<sup>+</sup> and DAPI<sup>+</sup> | Apoptotic   | PRJEB33078         | [ENA, BAM files](http://ftp.sra.ebi.ac.uk/vol1/run/ERR337/ERR3379692/apoptotic_genome.bam)       |
| * | HEK293 cell line  | Staurosporine (1 μM, 2 hours)      | Annexin V<sup>+</sup> and DAPI<sup>-</sup> | Pro-apoptotic | PRJEB33078         | [ENA, BAM files](http://ftp.sra.ebi.ac.uk/vol1/run/ERR337/ERR3379695/proapoptotic_genome.bam)     |
|   | HEK293 cell line  | -                                  | Annexin V<sup>-</sup> and DAPI<sup>-</sup> | Control     | PRJEB33078         | [ENA, BAM files](http://ftp.sra.ebi.ac.uk/vol1/run/ERR337/ERR3379693/healthy_genome.bam)       |
| | | | | | |||
| * | GM18507 cell line | TNF-alpha (100 ng/ml, 24 hours)    | Propidium iodide<sup>+</sup> and Annexin V<sup>+</sup>  | Dead        | EGAS00001003753         | [Zenodo, TENX049_SA928_003_sceset_v3_raw.rds](https://zenodo.org/records/3407791)      |
| * | GM18507 cell line | TNF-alpha (100 ng/ml, 24 hours)    | Propidium iodide<sup>-</sup> and Annexin V<sup>+</sup>  | Dying       | EGAS00001003753         | [Zenodo, TENX049_SA928_002_sceset_v3_raw.rds](https://zenodo.org/records/3407791)     |
|   | GM18507 cell line | -                                  | Propidium iodide<sup>-</sup> and Annexin V<sup>-</sup> | Control     | EGAS00001003753         | [Zenodo, TENX049_SA928_001_sceset_v3_raw.rds](https://zenodo.org/records/3407791)       |
| | | | | | |||
| * | PDX               | TNF-alpha (100 ng/ml, 24 hours)    | Propidium iodide<sup>+</sup> and Annexin V<sup>+</sup>  | Dead        | EGAS00001003753         | [Zenodo, TENX019_SA604X7XB02089_004_sceset_v3_raw.rds](https://zenodo.org/records/3407791)       |
|   | PDX               | -    | Propidium iodide<sup>-</sup> and Annexin V<sup>-</sup>  | Control       | EGAS00001003753         | [Zenodo, TENX019_SA604X7XB02089_002_sceset_v3_raw.rds](https://zenodo.org/records/3407791)       |


<br>
<br>

#### Non-ground truth data    |    15 datasets, 15 samples

| Dataset |  Tissue Type | Accession | Download | Data type | Protocol |
|:--------|:--------|:--------|:--------|:--------|:--------|
| _Healthy Human tissue extracts_ | | | | | | 
| 4K PBMC (10x Genomics 2017)                            | PBMC               | GSE108444 | [Link](https://s3-us-west-2.amazonaws.com/10x.files/samples/cell-exp/2.1.0/pbmc4k/pbmc4k_fastqs.tar)                                                                                                          | FASTQ        | 10x v2, normal         |
| Healthy Human Lung (Habermann, 2020)                   | Lung               | GSE135893 | [Link1](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR998/002/SRR9985352/SRR9985352_1.fastq.gz), [Link2](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR998/002/SRR9985352/SRR9985352_2.fastq.gz)                              | FASTQ        | 10x v2, extra length   |
| Liver atlas (MacParland, 2018)                         | Liver              | GSE115469 | [Link](http://ftp.sra.ebi.ac.uk/vol1/run/SRR727/SRR7276474/bam/P1TLH.bam)                                                                                                                                      | BAM to FASTQ | 10x v2, normal         |
| | | | | |
| _Diseased Human tissue extracts_ | | | | |
| TB PBMC (Cai, 2020)                                    | PBMC                | PRJNA605083 | [Link1](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR110/091/SRR11038991/SRR11038991_1.fastq.gz), [Link2](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR110/091/SRR11038991/SRR11038991_2.fastq.gz)                          | FASTQ        | 10x v2, normal         |
| TB Lung (Wang, 2023)                                   | Lung                | GSE192483 | [Link1](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR217/019/SRR21771719/SRR21771719_1.fastq.gz), [Link2](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR217/019/SRR21771719/SRR21771719_2.fastq.gz)                          | FASTQ        | 10x v2, extra length   |
| HBV Liver (Beudeker, 2024)                             | Liver               | GSE247322 | [Link](https://sra-pub-run-odp.s3.amazonaws.com/sra/SRR26730867/SRR26730867)                                                                                                                                   | SRA to FASTQ | 10x v3                 |
| | | | | |
| _Mouse tissue extracts_ | | | | |
| 10K Mouse PBMC (10X Genomics, 2021)                    | PBMC                 | [10X Genomics](https://www.10xgenomics.com/datasets/10-k-mouse-pbm-cs-multiplexed-2-cm-os-3-1-standard-6-0-0) | [Link](https://s3-us-west-2.amazonaws.com/10x.files/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_CellPlex_Mouse_PBMC_10K_Multiplex/SC3_v3_NextGem_DI_CellPlex_Mouse_PBMC_10K_Multiplex_fastqs.tar)               | FASTQ        | 10x v3                 |
| Mouse lung (Li, 2021)                                  | Lung                 | [10X Genomics](https://www.ncbi.nlm.nih.gov/sra/?term=SRR16190628) | [Link1](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR161/028/SRR16190628/SRR16190628_1.fastq.gz), [Link2](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR161/028/SRR16190628/SRR16190628_2.fastq.gz)                          | FASTQ        | 10x v3                 |
| Liver necroposis (Vucur, 2023)                         | Liver                | GSE166875 | [Link1](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR137/006/SRR13724006/SRR13724006_1.fastq.gz), [Link2](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR137/006/SRR13724006/SRR13724006_2.fastq.gz)                          | FASTQ        | 10x v2, normal         |
| | | | | |
| _Human tumours_ | | | | |
| Human Hodgkin's Lymphoma (10x Genomics 2020b)          | Lymphoid tissue      | [10X Genomics](https://www.10xgenomics.com/datasets/hodgkins-lymphoma-dissociated-tumor-targeted-immunology-panel-3-1-standard-4-0-0) | [Link](https://s3-us-west-2.amazonaws.com/10x.files/samples/cell-exp/4.0.0/Parent_NGSC3_DI_HodgkinsLymphoma/Parent_NGSC3_DI_HodgkinsLymphoma_fastqs.tar)                                                     | FASTQ        | 10x v3                 |
| Human Glioblastoma (10x Genomics 2020)                 | Glioblastoma         | [10X Genomics](https://www.10xgenomics.com/datasets/human-glioblastoma-multiforme-3-v-3-whole-transcriptome-analysis-3-standard-4-0-0) | [Link](https://cf.10xgenomics.com/samples/cell-exp/4.0.0/Parent_SC3v3_Human_Glioblastoma/Parent_SC3v3_Human_Glioblastoma_fastqs.tar)                                                                         | FASTQ        | 10x v3                 |
| Human Ductal Carcinoma (10x Genomics 2021)             | Breast cancer tissue | [10X Genomics](https://www.10xgenomics.com/datasets/7-5-k-sorted-cells-from-human-invasive-ductal-carcinoma-3-v-3-1-3-1-standard-6-0-0)  | [Link](https://s3-us-west-2.amazonaws.com/10x.files/samples/cell-exp/6.0.0/Breast_Cancer_3p/Breast_Cancer_3p_fastqs.tar)                                                                                     | FASTQ        | 10x v3                 |
| | | | | |
| _Human cell lines_ | | | | |
| 5k A549, Lung Carcinoma Cells (10X Genomics 2021)      | -                    | GSE161363 | [Link](https://s3-us-west-2.amazonaws.com/10x.files/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_CRISPR_A549_5K_Multiplex/SC3_v3_NextGem_DI_CRISPR_A549_5K_Multiplex_fastqs.tar)                                 | FASTQ        | 10x v3                 |
| Raji and Jurkat Cells (10X Genomics, 2021)             | -                    | [10X Genomics](https://www.10xgenomics.com/datasets/10-k-1-1-mixture-of-raji-and-jurkat-cells-multiplexed-2-cm-os-3-1-standard-6-0-0) | [Link](https://s3-us-west-2.amazonaws.com/10x.files/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex_fastqs.tar)             | FASTQ        | 10x v3                 |
| HCT-116 (Meir, 2016)                                   | -                    | GSE144357 | [Link1](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR109/069/SRR10974769/SRR10974769_1.fastq.gz), [Link2](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR109/069/SRR10974769/SRR10974769_2.fastq.gz)                          | FASTQ        | 10x v2, normal         |

<br>
<br>

### **Simulated data** 
_1 parent dataset, 30 final samples_

All simulated samples were created from a dataset available in ```R``` through the ```SeuratData``` package. The interferon-treated PBMC dataset was selected as it contains case and control data already intergrated into a single object with cell annotations. 

```R
# IFNB-Stimulated & Control PBMC data
library(SeuratData)
InstallData("ifnb")
data("ifnb")
ifnb <- UpdateSeuratObject(ifnb)
```

These foundational simulated samples were altered by randomly extracting a fixed percentage of their cells (2.5, 5, 10, 15, or 20), passing them through a damage-pertubration function, and reintroducing them to the sample. These are available for download on ```Zenodo```, [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.14048075.svg)](https://doi.org/10.5281/zenodo.14048075).

<br>

<div style="text-align: center;">
  <img src="https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/images/workflow_C.png" alt="Alt text" width="600"/>
</div>


<br>
<br>


## ⚪ Analysis Scripts
Details for running each component of the analysis are outlined in the scripts below 
<br>


### A - Searches 
* Survey of damaged strategies used in published scRNA-seq protocols over the past decade [analysis script](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/A%20-%20Searches/1%20-%20Manual%20Strategies%20Search.R), [pie chart plotting script](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/A%20-%20Searches/2%20-%20Manual%20Strategies%20Plot.R)
* Review of scRNA-seq tools capable of damaged cell detection [analysis and scatter plot script](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/A%20-%20Searches/3%20-%20Tool-based%20Search%20and%20Plot.R)


<br>

### B - Data preparation 

* Steps to convert raw data into the correct format, [SRA -> FASTQ](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/B%20-%20Processing%20raw%20data/i%20-%20SRA%20to%20FASTQ%20conversion.sh) or [BAM -> FASTQ](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/B%20-%20Processing%20raw%20data/ii%20-%20BAMFASTQ%20conversion.sh), before alignment and quantification with [STARsolo](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/B%20-%20Processing%20raw%20data/2%20-%20Align%20and%20quantify.sh) using genome indices generated with [STAR](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/B%20-%20Processing%20raw%20data/1%20-%20Generate%20genome%20index.sh).

<br>


### C - Running damaged strategies 
* Steps to create simulated data, including [fitting a model](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Simulation%20analyses%20/1.1.1%20-%20Fit%20scDesign2%20models.R) from parent data, [simulating](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Simulation%20analyses%20/1.1.2%20-%20Simulate%20datasets.R) daughter data using the model, and [perturbing]((https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Simulation%20analyses%20/1.2%20-%20Perturbation%20function.R). ) a fixed percentage of cells from each daughter dataset.
* Steps to [process](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Testing%20strategies%20on%20processed%20data/1.1%20-%20Pre-process%20raw%20data.R) existing datasets and make necessary adjustments to [groundtruth](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Testing%20strategies%20on%20processed%20data/1.2%20-%20Prepare%20ground%20truth%20samples.R) datasets.
* Running the strategies themselves requires separate scripts for the ```Python```-based strategies, [ddqc](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Testing%20strategies%20on%20processed%20data/2.1%20-%20Run%20ddqc.ipynb) and [EnsembleKQC](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Testing%20strategies%20on%20processed%20data/2.2%20-%20Run%20EnsembleKQC.sh), that must be run before collective testing on [```R```-based strategies](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Testing%20strategies%20on%20processed%20data/3%20-%20Run%20all%20damaged%20detection%20strategies%20on%20all%20data.R), using a function that requires three helper scripts, [1](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Testing%20strategies%20on%20processed%20data/3.1%20-%20Helper%20plotting%20function.R), [2](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Testing%20strategies%20on%20processed%20data/3.2%20-%20Helper%20summarising%20function.R), [3](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Testing%20strategies%20on%20processed%20data/3.3%20-%20Helper%20valiDrops%20function.R).
* Next, the performance of the strategies on [groundtruth datasets](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Testing%20strategies%20on%20processed%20data/4%20-%20Performance%20metrics%20for%20ground%20truth%20cases.R) is measured, as well as on simulated datasets, [1](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Simulation%20analyses%20/3.1%20-%20HVG%20similarity%20tests.R) and [2](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Simulation%20analyses%20/3.2%20-%20DGEA%20correctness%20tests.R). The results for these analyses are plotted within the code for the groundtruth data and separately [here](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Simulation%20analyses%20/4%20-%20Performance%20plots.R) for the simulated data.
* The output of the tools across the groundtruth, non-ground truth, and simulated data is then [calculated and plotted](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/C%20-%20Testing%20strategies%20on%20processed%20data/5%20-%20Comparison%20metrics%20for%20all%20cases.R).

<br>

### D. Recommentations
* Tools are ranked by performance, stability, and usability in this [script](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/D%20-%20Summarising%20results/1%20-%20Comparison%20metrics%20for%20all%20cases.R) to aid users selection. 
* Further investigation on the characteristics of the ground truth damaged cells is conducted [here](https://github.com/AlicenJoyHenning/DamageToolReviewArticle/blob/main/code/D%20-%20Isolated%20damaged%20cell%20analyses/1%20-%20Annotate%20and%20visualize%20damaged%20clusters.R) to inform developers on the shortcomings of existing strategies.

<br>


## Authors 

**Alicen Henning** <br>
alicen.jhb@gmail.com <br>
Stellenbosch University Tygerberg Medical Campus <br>

<br>

**Project supervisors** <br>
**Prof Marlo Möller** <br>
Stellenbosch University Tygerberg Medical Campus <br>

<br>

**Prof Andre Loxton** <br> 
Stellenbosch University Tygerberg Medical Campus <br>

<br> 

**Prof Gian van der Spuy** <br>
Stellenbosch University Tygerberg Medical Campus <br>


---


